_posts/2023-04-14-kafka-transactions.md

```diff
@@ -0,0 +1,304 @@
+
+```sh
+$ kafka-acls.sh --bootstrap-server:9092 --command-config client.properties --add \
+  --allow-principal User:* --operation write --transactional-id *
+```
+
+If any of the send operations inside the transaction scope fails, the final `commitTransaction` will fail and throw the exception from the last failed send.
+When this happens, your application should call `abortTransaction` to reset the state and apply a retry strategy.
+Some errors cannot be recovered and the application should terminate (`ProducerFencedException`, `FencedInstanceIdException`, `OutOfOrderSequenceException`).
+
+#### Monitoring
+
+When running in production, it is important to monitor the health of your Kafka deployments to maintain reliable performance.
+Some of the metrics that Kafka provides can be useful when dealing with transaction issues and tuning.
+
+On the broker side, you should should collect at least the count and 99 percentile for the following APIs.
+
+```sh
+# requests: FindCoordinator, InitProducerId, AddPartitionsToTxn, Produce, EndTxn
+kafka.network:type=RequestMetrics,name=RequestsPerSec,request=([\w]+)
+kafka.network:type=RequestMetrics,name=RequestQueueTimeMs,request=([\w]+)
+kafka.network:type=RequestMetrics,name=TotalTimeMs,request=([\w]+)
+kafka.network:type=RequestMetrics,name=LocalTimeMs,request=([\w]+)
+kafka.network:type=RequestMetrics,name=RemoteTimeMs,request=([\w]+)
+kafka.network:type=RequestMetrics,name=ThrottleTimeMs,request=([\w]+)
+kafka.network:type=RequestMetrics,name=ResponseQueueTimeMs,request=([\w]+)
+kafka.network:type=RequestMetrics,name=ResponseSendTimeMs,request=([\w]+)
+attributes:Count,99thPercentile
+```
+
+Producer metrics are more difficult to collect, but they help when broker logs reveal some misbehavior.
+
+```sh
+kafka.producer:type=producer-metrics,client-id=([-.\w]+)
+attributes:txn-init-time-ns-total,txn-begin-time-ns-total,txn-send-offsets-time-ns-total,txn-commit-time-ns-total,txn-abort-time-ns-total
+```
+
+### Drawbacks
+
+There are a few drawbacks when using Kafka transactions that developers should be aware of.
+
+Firstly, transactions can add some latency to an application due to the write amplification required to ensure EOS.
+While this may not be a significant issue for some applications, those that require real-time processing may be affected.
+
+For each transaction we have the following overhead, which is independent of the number of messages: 
+
+- some RPCs to register the partitions with the coordinator (can be batched)
+- one marker has to be written to each partition when completing a transaction
+- few state change records are appended to the internal transaction state log
+
+Most of this latency is on the producer side, where the transactions API is implemented.
+The performance degradation can be significant when having short transaction commit intervals.
+Increasing the transaction duration also increases the end-to-end latency, so it's a matter of tradeoff.
```

Comment:
Plus the risk an issue is hit and the whole transaction has to be aborted.

Reply:
Yes, it may also require additional tuning.